include/CLI/Validators.hpp

```diff
@@ -287,34 +292,39 @@ class IsMember : public Validator {
         : IsMember(std::vector<T>(values), std::forward<Args>(args)...) {}
 
     /// This checks to see if an item is in a set (empty function)
-    template <typename T>
-    explicit IsMember(T set)
-        : IsMember(std::move(set),
-                   std::function<typename IsMemberType<typename element_value_type<T>::type>::type(
-                       typename IsMemberType<typename element_value_type<T>::type>::type)>()) {}
+    template <typename T> explicit IsMember(T set) : IsMember(std::move(set), nullptr) {}
 
     /// This checks to see if an item is in a set: pointer or copy version. You can pass in a function that will filter
     /// both sides of the comparison before computing the comparison.
     template <typename T, typename F> explicit IsMember(T set, F filter_function) {
+
         // Get the type of the contained item - requires a container have ::value_type
-        using item_t = typename element_value_type<T>::type;
-        using local_item_t = typename IsMemberType<item_t>::type;
+        // if the type does not have first_type and second_type, these are both value_type
+        using element_t = typename detail::element_type<T>::type;            // Removes (smart) pointers if needed
+        using item_t = typename detail::pair_adaptor<element_t>::first_type; // Is value_type if not a map
+
+        using local_item_t = typename IsMemberType<item_t>::type; // This will convert bad types to good ones
+                                                                  // (const char * to std::string)
 
         // Make a local copy of the filter function, using a std::function if not one already
         std::function<local_item_t(local_item_t)> filter_fn = filter_function;
 
         // This is the type name for help, it will take the current version of the set contents
         tname_function = [set]() {
             std::stringstream out;
-            out << detail::type_name<item_t>() << " in {" << detail::join(detail::smart_deref(set), ",") << "}";
+            out << "{";
+            int i = 0; // I don't like counters like this
+            for(const auto &v : detail::smart_deref(set))
+                out << (i++ == 0 ? "" : ",") << detail::pair_adaptor<element_t>::first(v);
+            out << "}";
```

Comment:
In one of the PR's I made a join overload that could take a lambda, that would get rid of the counter.

Reply:
Noted, will try to switch to this once that is merged.